src/lib/textToEdit.ts

```diff
@@ -0,0 +1,243 @@
+import { Models } from '@kittycad/lib'
+import { VITE_KC_API_BASE_URL } from 'env'
+import crossPlatformFetch from './crossPlatformFetch'
+import { err, reportRejection } from './trap'
+import { toSync } from './utils'
+import { Selections } from './selections'
+import { ArtifactGraph, getArtifactOfTypes } from 'lang/std/artifactGraph'
+import { SourceRange } from 'lang/wasm'
+
+function sourceIndexToLineColumn(
+  code: string,
+  index: number
+): { line: number; column: number } {
+  const codeStart = code.slice(0, index)
+  const lines = codeStart.split('\n')
+  const line = lines.length
+  const column = lines[lines.length - 1].length
+  return { line, column }
+}
+
+function convertAppRangeToApiRange(
+  range: SourceRange,
+  code: string
+): Models['SourceRange_type'] {
+  return {
+    start: sourceIndexToLineColumn(code, range[0]),
+    end: sourceIndexToLineColumn(code, range[1]),
+  }
+}
+
+export async function submitPromptToEditToQueue({
+  prompt,
+  selections,
+  code,
+  token,
+  artifactGraph,
+}: {
+  prompt: string
+  selections: Selections
+  code: string
+  token?: string
+  artifactGraph: ArtifactGraph
+}): Promise<Models['TextToCadIteration_type'] | Error> {
+  const ranges: Models['TextToCadIterationBody_type']['source_ranges'] =
+    selections.graphSelections.flatMap((selection) => {
+      const artifact = selection.artifact
+      const prompts: Models['TextToCadIterationBody_type']['source_ranges'] = []
+
+      if (artifact?.type === 'cap') {
+        prompts.push({
+          prompt: `The users main selection is the end cap of a general-sweep (that is an extrusion, revolve, sweep or loft).
+The source range most likely refers to "startProfileAt" simply because this is the start of the profile that was swept.
+If you need to operate on this cap, for example for sketching on the face, you can use the special string ${
+            artifact.subType === 'end' ? 'END' : 'START'
+          } i.e. \`startSketchOn(someSweepVariable, ${
+            artifact.subType === 'end' ? 'END' : 'START'
+          })\`
+When they made this selection they main have intended this surface directly or meant something more broad like the sweep itself.
+See later source ranges for more context.`,
+          range: convertAppRangeToApiRange(selection.codeRef.range, code),
```

Comment:
You can there's lots of context given to the model, both on what has been selected, but also what the user might have intended by the selection i.e.

They selected the face, but they might mean the whole extrusion.

This has proven useful and made some user prompts succeed when they previously failed 

@greg-kcio since this seems to be working, thinking a bit down the road, we probably want to bench mark these prompts too right? to make sure we're always improving, not sure how best to do that since some stuff would need to be shared between the FE and ML teams code. 

Just something to think about.

Reply:
How about a similar test setup to Text-to-CAD? Accuracy and performance benchmarks for a variety of prompts are run on a daily schedule, entirely in the backend via test harness for the server. The FE tests use simpler prompts and are only responsible for testing the end-to-end connection.

Reply:
I didn't fully understand how you were providing context when I wrote the previous comment.

Initially I thought that any prompt templating would happen only in the backend, but I see the benefit of using the artifact graph. This changes my thoughts above, a little bit. We could include these templates in the mock API calls used by the backend test harness. That way, FE tests would be responsible for testing that the right templates are used each selection type. End-to-end tests in the FE would still be good.

We can also experiment with moving some or all of your prompt templating into the backend.

Reply:
Yes!